components/style/properties/properties.mako.rs

```diff
@@ -3085,9 +3090,338 @@ pub fn cascade(
+                    .and_then(|registration| registration.initial_value.clone())
+                    .map(|initial_value| initial_value.to_computed_value(context).unwrap());
+                if let Some(initial_typed_value) = initial_typed_value {
+                    let initial_untyped_value = (&initial_typed_value).into();
+                    computed.insert_typed(&name, Some(initial_untyped_value), initial_typed_value);
+                }
+            }
+
+            substitute_all(
+                context,
+                &*registered_property_set,
+                specified,
+                inherited,
+                /* to_substitute */ &Some(dependencies),
+                &mut computed
+            );
+
+            (Some(Arc::new(computed)), font_size_cyclical)
+        } else if let &Some(ref inherited) = inherited {
+            // Nothing specified, but we did inherit something.
+
+            if inherited.has_uninherited() {
+                // There were uninherited custom properties set to something
+                // other than their initial value, so we have to create a new
+                // CustomPropertiesMap with them removed.
+                let mut computed = CustomPropertiesMap::new();
+                for (name, value) in inherited.iter() {
+                    let registration = registered_property_set.get(name);
+                    if registration.map(|x| x.inherits).unwrap_or(true) {
+                        // This property inherits; just clone it over.
+                        computed.insert(name.clone(), value.clone());
+                        if let Some(typed_value) = inherited.get_typed(name) {
+                            computed.insert_typed(name, None, typed_value.clone());
+                        }
+                    } else {
+                        // The property doesn't inherit; set it to its initial
+                        // value.
+                        if let Some(ref initial) = registration.and_then(|x| x.initial_value.as_ref()) {
+                            let typed_value = initial.clone().to_computed_value(context).unwrap();
+                            let untyped_value = (&typed_value).into();
+                            computed.insert_typed(name, Some(untyped_value), typed_value);
+                        }
+                    }
+                }
+                (Some(Arc::new(computed)), false)
+            } else {
+                // There weren't any uninherited custom properties set to
+                // something other than their initial value, so we can just
+                // clone the Arc, which should be cheap!
+                (Some((*inherited).clone()), false)
+            }
+        } else {
+            // There were no inherited or specified values.
+            // If there are any properties with initial values, we should set
+            // them.
+            (Some(Arc::new(registered_property_set.initial_values_except(context, None))), false)
```

Comment:
So we're creating this and calling `initial_values_except` for every element, even if we don't use custom properties?

That seems quite suboptimal, not only for the unnecessary allocation specially since the time `initial_values_except` takes grows with the number of registered properties :/

Also, no branch seems to return `None` for `computed_early`, is it intentional? If not, something needs to get fixed. If it is, I don't see why it should return an `Option`.

Reply:
> So we're creating this and calling initial_values_except for every element, even if we don't use custom properties?
> That seems quite suboptimal, not only for the unnecessary allocation specially since the time initial_values_except takes grows with the number of registered properties :/

Well, creating the Arc is expensive. But the cost of `initial_values_except` would not apply in the case you mentioned except on the root, as either we are inheriting precisely these values, or in fact there are no registered properties whatsoever.

I will fix it so that if there are no registered properties we return None as we would have in the past. In the future, and in line with https://github.com/w3c/css-houdini-drafts/issues/286, it seems we should really be not storing computed initial values here at all (so that we can be completely analogous with regular properties); will have to make sure to do in a nice way so that we don't violate the abstraction barrier between substition and registrations.